fix(cli): do not create `bin` folder within `DENO_INSTALL_ROOT`

mahtaran opened the pull request:
As can be seen in the documentation, `DENO_INSTALL_ROOT` is meant to be the folder wherein Deno will install executables, with the default being `$HOME/.deno/bin`.

However, the current behaviour of the `install` and `uninstall` commands, if `DENO_INSTALL_ROOT` is set, is to still append "bin" to the path. This leads to executables being installed in a folder like `$HOME/.deno/bin/bin`, which is obviously not the intended behaviour.

This commit removes this appending. Do note that this is a breaking change in the behaviour of the `--root` flag, as previously it referred to a folder wherein the `bin` folder was located, but now it refers to the `bin` folder itself. For example: `deno install --root /foo/bar test.ts` would previously install the executable in `/foo/bar/bin/test`, but now it will install it in `/foo/bar/test`.

This would also most certainly break many third-party tools that expect the current behaviour. See for example:
- https://github.com/jdx/mise/blob/0cea93b27bc46ba9219557532d6d1d061d05f81d/src/plugins/core/deno.rs#L144
- https://github.com/asdf-community/asdf-deno/blob/13f7f08c93989fdc0b372e78df226e511fc13fa9/bin/exec-env#L7

Closes #26442 

Please do note that, as I am currently on a family visit out of the country, I currently only have my poor little laptop. As such, I do not have the storage capacity to build the project or run the test suite, so **THIS CODE IS UNTESTED**. If someone else could test it for me, that'd be grand, and otherwise I'll do it when I'm back home in a couple of weeks.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/denoland/deno?pullRequest=27459) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/denoland/deno?pullRequest=27459) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/denoland/deno?pullRequest=27459) it.</sub>

mahtaran commented:
I seem to have completely missed that @dovakin0007 already submitted a fix in #26446, apologies! I do believe this PR is more 'correct' as it modifies the behaviour to be what the documentation says it is, however Gowtham's PR will cause noticeably less (but likely still some) breaking of third-party tools and existing set-ups. I'll leave the choice on what to do up to the maintainers.

mahtaran committed:
fix(cli): do not create `bin` folder within `DENO_INSTALL_ROOT`

As can be seen in the documentation, `DENO_INSTALL_ROOT` is meant to be the folder wherein Deno will install executables, with the default being `$HOME/.deno/bin`.

However, the current behaviour of the `install` and `uninstall` commands, if `DENO_INSTALL_ROOT` is set, is to still append "bin" to the path. This leads to executables being installed in a folder like `$HOME/.deno/bin/bin`, which is obviously not the intended behaviour.

This commit removes this appending. Do note that this is a breaking change in the behaviour of the `--root` flag, as previously it referred to a folder wherein the `bin` folder was located, but now it refers to the `bin` folder itself. For example: `deno install --root /foo/bar test.ts` would previously install the executable in `/foo/bar/bin/test`, but now it will install it in `/foo/bar/test`.

mahtaran commented:
Alright, that took a _little_ longer than planned, but I just built this out on my PC and ran the test suite, and everything seems to be working well! I do believe this should now be ready for review, please let me know if anything needs be changed.

dsherret committed:
Merge branch 'main' into fix/install-root

dsherret committed:
updates based on main

dsherret committed:
revert cargo.lock changes

dsherret reviewed (commented):
It looks good to me, but I'm slightly hesitant to merge this because it's a breaking change for probably undesired behaviour. I'll raise it internally.

bartlomieju commented:
Hi @mahtaran, thank you for the PR. We've discussed it internally during the team meeting and we are on the fence, leaning towards not merging it. The fact that changing this behavior will have a breaking effect on 3rd party programs (like `asdf`) makes us believe that it's not worth fixing it, given that this is a niche feature. Would documenting the bahavior of this env var be satisfactory to you?

bartlomieju commented:
Thank you for the PR and sorry it sat for so long. We've decided not to change the `DENO_INSTALL_ROOT` semantics: the worst symptom from #26442 (installing to `$HOME/.deno/bin/bin` when the env var is set to the documented default) was fixed in #26446, which skips appending `bin` when the path already ends with `bin`. With that guard in place both `$HOME/.deno` and `$HOME/.deno/bin` resolve to the same install directory.

Changing the env var (and `--root`) to be used verbatim would silently break version managers like mise and asdf-deno, as well as dotfiles and CI scripts that set the root to a parent directory and put `$DENO_INSTALL_ROOT/bin` on `PATH` - as you correctly pointed out in the description. The remaining inconsistency was on the docs side, which is addressed in denoland/docs#3381.

mahtaran commented:
> Thank you for the PR and sorry it sat for so long. We've decided not to change the `DENO_INSTALL_ROOT` semantics: the worst symptom from #26442 (installing to `$HOME/.deno/bin/bin` when the env var is set to the documented default) was fixed in #26446, which skips appending `bin` when the path already ends with `bin`. With that guard in place both `$HOME/.deno` and `$HOME/.deno/bin` resolve to the same install directory.
> 
> Changing the env var (and `--root`) to be used verbatim would silently break version managers like mise and asdf-deno, as well as dotfiles and CI scripts that set the root to a parent directory and put `$DENO_INSTALL_ROOT/bin` on `PATH` - as you correctly pointed out in the description. The remaining inconsistency was on the docs side, which is addressed in denoland/docs#3381.

That seems like a sensible fix, thank you very much for addressing it!
